atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueService.java

```diff
@@ -1345,50 +1344,87 @@ private CfDef getCfForTable(TableReference tableRef, byte[] rawMetadata) {
         final SlicePredicate pred = new SlicePredicate();
         pred.setSlice_range(slice);
 
-        final ColumnSelection selection = rangeRequest.getColumnNames().isEmpty() ? ColumnSelection.all()
-                : ColumnSelection.create(rangeRequest.getColumnNames());
 
         AbstractPagingIterable<RowResult<U>, TokenBackedBasicResultsPage<RowResult<U>, byte[]>> rowResults =
-                new AbstractPagingIterable<RowResult<U>, TokenBackedBasicResultsPage<RowResult<U>, byte[]>>() {
-            private final ColumnGetter columnGetter = new CqlColumnGetter(consistency, tableRef, columnBatchSize);
+                new PagingIterable(
+                        new CqlColumnGetter(consistency, tableRef, columnBatchSize),
+                        rangeRequest,
+                        tableRef,
+                        pred,
+                        consistency,
+                        resultsExtractor,
+                        timestamp
+                );
 
-            @Override
-            protected TokenBackedBasicResultsPage<RowResult<U>, byte[]> getFirstPage() throws Exception {
-                return getSinglePage(rangeRequest.getStartInclusive());
-            }
+        return ClosableIterators.wrap(rowResults.iterator());
+    }
 
-            @Override
-            protected TokenBackedBasicResultsPage<RowResult<U>, byte[]> getNextPage(
-                    TokenBackedBasicResultsPage<RowResult<U>, byte[]> previous) throws Exception {
-                return getSinglePage(previous.getTokenForNextPage());
-            }
+    private class PagingIterable<U, T> extends AbstractPagingIterable {
```

Comment:
Yep, there are still a few odd things.  I also want to get rid of the cast in `getNextPage` and move these classes to a separate file.  Had a few other can't-slip deadlines tonight though, so wanted to push WIP just in case.